File: app/models/achievement.py
from sqlalchemy import Column, Integer, String, ForeignKey, DateTime, Enum, Text
from sqlalchemy.sql import func
from sqlalchemy.orm import relationship

from app.infrastructure.database import Base
from app.models.enums import AchievementStatus, AchievementCategory, AchievementLevel


class Achievement(Base):
    __tablename__ = "achievements"

    id = Column(Integer, primary_key=True, index=True)
    user_id = Column(Integer, ForeignKey("users.id", ondelete="CASCADE"), nullable=False)

    title = Column(String, nullable=False)
    description = Column(Text, nullable=True)
    file_path = Column(String, nullable=False)

    category = Column(Enum(AchievementCategory), default=AchievementCategory.OTHER, nullable=False)
    level = Column(Enum(AchievementLevel), default=AchievementLevel.SCHOOL, nullable=False)
    points = Column(Integer, default=0)  # Баллы за достижение

    status = Column(Enum(AchievementStatus), default=AchievementStatus.PENDING)
    rejection_reason = Column(Text, nullable=True)

    created_at = Column(DateTime(timezone=True), default=func.now(), server_default=func.now())
    updated_at = Column(DateTime(timezone=True), default=func.now(), onupdate=func.now())

    user = relationship("Users", back_populates="achievements")

File: frontend/src/types/achievement.ts
import {
  AchievementStatus,
  AchievementCategory,
  AchievementLevel,
  AchievementResult,
} from './enums'

export interface Achievement {
  id: number
  user_id: number
  title: string
  description?: string
  file_path?: string | null
  external_url?: string | null
  category: AchievementCategory
  level: AchievementLevel
  result?: AchievementResult
  points: number
  status: AchievementStatus
  rejection_reason?: string
  moderator_id?: number
  created_at: string
  updated_at: string
  user?: {
    id: number
    first_name: string
    last_name: string
    email: string
    education_level?: string
  }
  projected_points?: number
}

export interface AchievementListResponse {
  achievements: Achievement[]
  page: number
  total_pages: number
}

export interface ModerationAchievementsResponse {
  achievements: Achievement[]
  stats: {
    total_pending: number
  }
  page: number
  total_pages: number
}
